File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.py
def solution(a, b, c, d):
    l1 = [a, b, c, d]
    l2 = list(set(l1)) 
    if len(l2) == 1:
        return 1111 * l2[0]
    if len(l2) == 2:
        cnt = l1.count(l2[0])
        if cnt == 1:
            return (10 * l2[1] + l2[0])**2
        elif cnt == 2:
            return (l2[0] + l2[1]) * abs(l2[0] - l2[1])
        elif cnt == 3:
            return (10 * l2[0] + l2[1]) ** 2
    if len(l2) == 3:
        for i in range(3):
            if l1.count(l2[i]) == 2:
                idx = i
        l2.pop(idx)
        return l2[0] * l2[1]
    if len(l2) == 4:
        return min(l1)

File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.cpp
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int solution(int a, int b, int c, int d) {
    vector<int> v = {a, b, c, d};
    sort(v.begin(), v.end());
    
    if (v[0] == v[3]) {
        return 1111 * v[0];
    }
    
    if (v[1] == v[3]) {
        int p = v[1];
        int q = v[0];
        return (10 * p + q) * (10 * p + q);
    }
    
    if (v[0] == v[2]) {
        int p = v[0];
        int q = v[3];
        return (10 * p + q) * (10 * p + q);
    }
    
    if (v[0] == v[1] && v[2] == v[3]) {
        int p = v[0];
        int q = v[2];
        return (p + q) * abs(p - q);
    }
    
    if (v[0] == v[1]) return v[2] * v[3];
    if (v[1] == v[2]) return v[0] * v[3];
    if (v[2] == v[3]) return v[0] * v[1];
    
    return v[0];
}